Add OCF Pacemaker RA for HA A/A cluster

bogdando opened the pull request:
Add OCF script for Pacemaker resource agent
for dynamic membership control of the nodes
in RabbitMQ active-active mirrored cluster.

source https://github.com/stackforge/fuel-library
v6.1-dev b9b6376cab317906af625faa5e48521aaeb0201b

Co-Author: Vladimir Kuklin <vkuklin@mirantis.com>
Co-Author: Sergey Vasilenko <svasilenko@mirantis.com>
Signed-off-by: Bogdan Dobrelya <bdobrelia@mirantis.com>

dumbbell commented:
Hi!

Thank you for your contribution! I'm going to post here the comments I previously wrote by email.

dumbbell commented on packaging/RPMS/Fedora/rabbitmq-server.spec:
```diff
@@ -11,6 +11,7 @@ Source2: rabbitmq-script-wrapper
 Source3: rabbitmq-server.logrotate
 Source4: rabbitmq-server.ocf
 Source5: README
+Source6: rabbitmq-server-ha.ocf
```
Please put `rabbitmq-server-ha.ocf` just below `rabbitmq-server.ocf`.

bogdando replied:
Do you mean inserting ocf as source5 and renaming all shifted source5 to source6 ? This would introduce more unneeded changes, that is the point?

bogdando replied:
This seems a cosmetic fix, although I agree with you. Let's keep this for the next change set please

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+#!/bin/bash
```
Why do you need Bash?

After looking at the code, the main Bash feature used is `[[ $var == "literal" ]]`. This is easily replaced with `[ "$var" = "literal" ]`.

bogdando replied:
We have much bashisms in this script, I suggest to address refactoring to the "sh" by separate changes. Hopefully once we would have tests and maybe CI for this script as well

bogdando replied:
@dumbbell : If we wanted to cover this script with bats unit tests, should we want to keep it written as "#!/bin/bash" or make it 100% POSIX compatible won't hurt?

michaelklishin replied:
@bogdando I'd focus on the platforms/shells you plan using this script on. We are not pedantic. That said, @dumbbell can provide you a lot of feedback on *BSD compatibility.

bogdando replied:
Understood. Please note, that OCF scripts in pacemaker seems always importing an ocf-shellfuncs "library", which has a bash shebang. Although, I have the version of this OCF script 100% /bin/sh compatible, I doubt it makes sense to stick with a "sh" shebang here because of the aforementioned ocf-shellfuncs... But I will update this patch set as well to be shell compatible :)

bogdando replied:
Done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+OCF_RESKEY_debug_default=false
+OCF_RESKEY_username_default="rabbitmq"
+OCF_RESKEY_groupname_default="rabbitmq"
+OCF_RESKEY_pid_file_default=/var/run/rabbitmq/p_pid
```
You could re-use the same default value as the current OCF RA.

bogdando replied:
will do

bogdando replied:
done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+rmq_setup_env() {
+    local H
+    local dir
+    H=`hostname -s`
```
This can only work with Erlang nodes using short names, not long names.

bogdando replied:
should I introduce an OCF parameter like use_fqdn?

michaelklishin replied:
Yes, please. In some environments, namely EC2, using long names are more convenient. RabbitMQ itself has an option for using long names these days.

bogdando replied:
done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+    local H
+    local dir
+    H=`hostname -s`
+    export RABBITMQ_NODENAME="rabbit@${H}"
```
You already have `rabbit_node_name` below to format a node name, you could use it here as well.

bogdando replied:
done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+}
+
+rabbit_node_name() {
+    echo "rabbit@"$(echo "$1" | awk -F. '{print $1}')
```
As stated earlier, this can only work with Erlang nodes using short names, not long names.

bogdando replied:
done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+    # remove mnesia files, if required
+    if $make_amnesia ; then
+        kill_rmq_and_remove_pid
+        ocf_run rm -rf ${MNESIA_FILES}*
```
If `$OCF_RESKEY_mnesia_base` contains data for several nodes, the above command may remove other nodes data, if they are named `rabbit@node1`, `rabbit@node10`, `rabbit@node11`, etc. if we are resetting `rabbit@node1`. Adding a slash will fix the problem.

bogdando replied:
Done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+    # otherwise, try to add a blocking rule with max of 5 retries
+    local tries=5
+    until $(iptables -nvL | grep -q 'temporary RMQ block') || [[ $tries -eq 0 ]]; do
+      ((tries--))
```
This is another bashism. You can replace it with `$((tries--))`.

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+    elif [ "$1" == 'running' ]
+    then
+        infotype='running_db_nodes'
+    fi
```
Style issue: this whole `if` block would be more readable as a `case`.

bogdando replied:
This seems a cosmetic fix, although I agree with you. Let's keep this for the next change set please

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+        return $OCF_ERR_GENERIC
+    fi
+    # translate line like '{running_nodes,['rabbit@node-1','rabbit@node-2','rabbit@node-3']},' to node_list
+    echo $(echo "${c_status}" | grep "${cl}" | awk -F, '{ for (i=1;i<=NF;i++) { if ($i ~ /@/) { gsub(/[\[\]}{]/,"",$i); print $i; } }}' | tr -d  "\'")
```
Minor comment, not just for this line: awk(1) can perform the regex matching, no need for an additional grep(1) in front of it. On other lines, the same comment applies to the combination of awk(1) and sed(1).

bogdando replied:
This seems a cosmetic fix, although I agree with you. Let's keep this for the next change set please

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+        fi
+        # todo: check content for digital
+        if [[ -d /proc/${pid}/ ]] ; then
+            ocf_run kill -9 $pid
```
`SIGTERM` is probably enough and less brutal.

bogdando replied:
Done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+            ocf_log warn "${LH} RMQ-runtime (beam) PID=${pid} stopped by 'kill -9', sorry..."
+        fi
+        ocf_run rm -f $OCF_RESKEY_pid_file
+    fi
```
Would it be worth checking for a running `beam` or `beam.smp` without the associated PID file? `beam` command line contains the node name.

bogdando replied:
sounds reasonable

bogdando replied:
Done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+    if [[ -f $OCF_RESKEY_pid_file ]] ; then
+        ocf_log warn "${LH} found old PID-file '${OCF_RESKEY_pid_file}'."
+        pid=$(cat ${OCF_RESKEY_pid_file})
+        if [[ -d /proc/${pid} && ! -z ${pid} ]] ; then
```
Style issue: `! -z $pid` could be simplified as just `"$pid"`.

Also, it would be more logical to test it before the directory.

bogdando replied:
Done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+            rc=$?
+            if [[ $rc == $OCF_SUCCESS ]] ; then
+                ocf_log warn "${LH} found beam process with PID=${pid}, killing...'."
+                ocf_run kill -9 $pid
```
Again, `SIGTERM` should be enough.

bogdando replied:
Done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+                return $OCF_ERR_GENERIC
+            fi
+        fi
+        ocf_run rm -rf $OCF_RESKEY_pid_file
```
`rm -f` is enough and safer in case of error/typo.

bogdando replied:
Done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+    RABBITMQ_NODE_ONLY=1 su rabbitmq -s /bin/sh -c "${command}"&
+    ts_end=$(( $(now) + ${OCF_RESKEY_start_time} ))
+    rc=$OCF_ERR_GENERIC
+    while [ $(now) -lt ${ts_end} ]; do
```
Can't you use `rabbitmqctl wait` for this purpose, instead of this loop?

bogdando replied:
will do

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+  #   if (EnabledPlugins > 0) and (ActivePlugins == 0) ; then it's safe to load
+  # If we have at least one active plugin, then it's not safe to re-load them
+  # because plugins:setup() would remove existing dependency plugins in plugins_expand_dir.
+  ${OCF_RESKEY_ctl} eval '{ok, EnabledFile} = application:get_env(rabbit, enabled_plugins_file), EnabledPlugins = rabbit_plugins:read_enabled(EnabledFile), ActivePlugins = rabbit_plugins:active(), if length(EnabledPlugins)>0 -> if length(ActivePlugins)==0 -> erlang:error("need_to_load_plugins"); true -> false end; true -> false end.'
```
Can't you use `rabbitmq-plugins` for this?

Should we add more options and output formats to the command?

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+}
+
+list_active_plugins() {
+  local LIST=`${OCF_RESKEY_ctl} eval 'rabbit_plugins:active().'`
```
Likewise, can't you use rabbitmq-plugins for this?

bogdando replied:
This seems quite a cosmetic change, if you don't mind I will skip this for now.

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+  if [[ $? == 0 ]] ; then
+    return 0
+  else
+    ${OCF_RESKEY_ctl} eval 'ToBeLoaded = rabbit_plugins:setup(), ok = app_utils:load_applications(ToBeLoaded), StartupApps = app_utils:app_dependency_order(ToBeLoaded,false), app_utils:start_applications(StartupApps).'
```
Can't you use `rabbitmq-plugins` for this?

Should we add more options and output formats to the command?

bogdando replied:
This seems quite a cosmetic change, if you don't mind I will skip this for now.

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+        echo "$d  [notify]  ${OCF_RESKEY_CRM_meta_notify_type}-${OCF_RESKEY_CRM_meta_notify_operation} promote='${OCF_RESKEY_CRM_meta_notify_promote_uname}' demote='${OCF_RESKEY_CRM_meta_notify_demote_uname}' master='${OCF_RESKEY_CRM_meta_notify_master_uname}' slave='${OCF_RESKEY_CRM_meta_notify_slave_uname}' start='${OCF_RESKEY_CRM_meta_notify_start_uname}' stop='${OCF_RESKEY_CRM_meta_notify_sto ...
+    fi
+
+     if [[ ${OCF_RESKEY_CRM_meta_notify_type} == 'pre' ]] ; then
```
5-column indentation, this is inconsistent with the rest of the file.

bogdando replied:
done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+                  return $OCF_NOT_RUNNING
+                fi
+                # Note, this should fail when the mnesia is inconsistent.
+                # For example, when the "old" master processing the promition of the new one.
```
Typo: promition -> promotion.

bogdando replied:
done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+                my_host "${OCF_RESKEY_CRM_meta_notify_stop_uname}"
+                rc=$?
+                if [[ $rc != $OCF_SUCCESS ]] ; then
+                    # On ohter nodes processing the post-stop, make sure the stopped node will be forgotten
```
Typo: ohter -> other.

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+                    rabbitmqctl set_policy ha-all "." '{"ha-mode":"all", "ha-sync-mode":"automatic"}' --apply-to all --priority 0
+                    rabbitmqctl set_policy heat_rpc_expire "^heat-engine-listener\\." '{"expires":3600000,"ha-mode":"all","ha-sync-mode":"automatic"}' --apply-to all --priority 1
+                    rabbitmqctl set_policy results_expire "^results\\." '{"expires":3600000,"ha-mode":"all","ha-sync-mode":"automatic"}' --apply-to all --priority 1
+                    rabbitmqctl set_policy tasks_expire "^tasks\\." '{"expires":3600000,"ha-mode":"all","ha-sync-mode":"automatic"}' --apply-to all --priority 1
```
You should use `$OCF_RESKEY_ctl` instead of hard-coding rabbitmqctl.

bogdando replied:
done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+                    get_monitor
+                    rc=$?
+                    ocf_log info "${LH} Master status is $rc"
+                    if [ $rc == $OCF_RUNNING_MASTER ]
```
The syntax of this test is invalid: you can't use `==` with `[`. It should be:
```sh
if [ "$rc" = "$OCF_RUNNING_MASTER" ]
```

dumbbell commented on packaging/debs/Debian/debian/rules:
```diff
@@ -19,6 +19,7 @@ install/rabbitmq-server::
 	done
 	sed -e 's|@RABBIT_LIB@|/usr/lib/rabbitmq/lib/rabbitmq_server-$(DEB_UPSTREAM_VERSION)|g' <debian/postrm.in >debian/postrm
 	install -p -D -m 0755 debian/rabbitmq-server.ocf $(DEB_DESTDIR)usr/lib/ocf/resource.d/rabbitmq/rabbitmq-server
+        install -p -D -m 0755 debian/rabbitmq-server-ha.ocf $(DEB_DESTDIR)usr/lib/ocf/resource.d/rabbitmq/rabbitmq-server-ha
```
You must use a tab to indent the line.

bogdando replied:
done

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+    local infotype=''
+    local rc=$OCF_ERR_GENERIC
+
+    if [ "$1" == 'nodes' ]
```
The syntax of this test is invalid: `[` doesn't support `==` as an operator, but see next comment.

dumbbell commented on packaging/common/rabbitmq-server-ha.ocf:
```diff
@@ -0,0 +1,1585 @@
+# Get alive cluster nodes in visible partition, but the specified one
+get_alive_pacemaker_nodes_but()
+{
+    if [ -z $1 ]; then
```
Please quote arguments to `[`. If the variable expands to the empty string, `[` will report a syntax error here.

This comment applies to the whole file.

bogdando commented:
@dumbbell thank you for a very thorough review and time spent. I will address all of the vital issues

bogdando commented:
@dumbbell, addressed your comments for the bash related things. The checkbashisms tool from the devscript, which I used to make this OCF script POSIX compatible, should not show warnings now.

bogdando commented:
@michaelklishin , @dumbbell addressed the pidfile and FQDN related comments.

bogdando commented:
Addressed the rest of the changes, but cosmetic ones. Let's please keep them to be fixed in next patch sets after the this one accepted. I believe in this huge file, there are *many* cosmetic things to be polished :)

dumbbell commented:
I agree, let's merge this first and address style changes later.
